plugins/action/common/prepare_plugins/prep_108_route_control.py

```diff
@@ -0,0 +1,297 @@
+# Copyright (c) 2024 Cisco Systems, Inc. and its affiliates
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of
+# this software and associated documentation files (the "Software"), to deal in
+# the Software without restriction, including without limitation the rights to
+# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+# the Software, and to permit persons to whom the Software is furnished to do so,
+# subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+# SPDX-License-Identifier: MIT
+
+import re
+from jinja2 import ChainableUndefined, Environment, FileSystemLoader
+from ansible_collections.ansible.utils.plugins.filter import ipaddr, hwaddr
+from ....plugin_utils.helper_functions import hostname_to_ip_mapping
+
+
+class PreparePlugin:
+    """
+    Class PreparePlugin
+    """
+    def __init__(self, **kwargs):
+        self.kwargs = kwargs
+        self.keys = []
+
+    def prepare(self):
+        """
+        function to prepare data for route_control
+        """
+        templates_path = self.kwargs['templates_path']
+        model_data = self.kwargs['results']['model_extended']
+        default_values = self.kwargs['default_values']
+
+        template_filename = "ndfc_route_control.j2"
+
+        env = Environment(
+            loader=FileSystemLoader(templates_path),
+            undefined=ChainableUndefined,
+            lstrip_blocks=True,
+            trim_blocks=True,
+        )
+
+        env.filters["ipaddr"] = ipaddr.ipaddr
+        env.filters["hwaddr"] = hwaddr.hwaddr
+        template = env.get_template(template_filename)
+
+        if "overlay_extensions" in model_data["vxlan"]:
```

Comment:
We should be able to significantly simplify this logic and remove the deep nesting by making use of data_model_key_check.

For example you can use it to walk the model_data tree with a set of keys and inspect to see if there is data at a particular level.  If you want to see if there is data under `["vxlan"]["overlay_extensions"]["route_control"]["route_maps"]` without all of the looping you could do this.

```python
from ....plugin_utils.helper_functions import data_model_key_check

parent_keys = ['vxlan', 'overlay_extensions', 'route_control', 'route_maps']
dm_check = data_model_key_check(model_data, parent_keys)
```

dm_check will contain the following dictionaries

```
{'keys_found': [], 'keys_not_found': [], 'keys_data': [], 'keys_no_data': []}
```

* 'keys_found' - indicates we found the keys in the model data but not necessarily values for those keys
* 'keys_not_foind' - indicates we did not find the keys in the model data
* 'keys_data' - indicates we found the keys and values/data associated with the keys
* 'keys_no_data' - indicates we found keys but no values/data associated with the keys

Reply:
Our prep_001 plugin has good examples of how its used to avoid deep nesting

https://github.com/netascode/ansible-dc-vxlan/blob/a9e6d181e79c430a21504aa1ad6fa53f1ef7a8dc/plugins/action/common/prepare_plugins/prep_001_list_defaults.py#L76

Reply:
Also might want to consider creating this function in the helper_functions library to safely pull data from the model

https://github.com/netascode/ansible-dc-vxlan/blob/a9e6d181e79c430a21504aa1ad6fa53f1ef7a8dc/roles/validate/files/rules/required_rules/401_overlay_services_cross_reference.py#L72